### PR TITLE
feat: Nats Jetstream consumer #11046 add simple support for jetstream subjects

### LINKED ISSUE #11046: Nats Jetstream consumer
## Feature Request

Nats.io Jetstream consumer

### Proposal:

The Nats server has a new streaming technology called jetstream. I work with nats using jetstream and it would be very helpful when Telegraf could consume messages from that technology https://docs.nats.io/nats-concepts/jetstream

### Current behavior:

Nats consumer plugin consumes data but not from a jetstream consumer

### Desired behavior:

Telegraf can be used to consume data from a nats jetstream consumer

### Use case: 
Nats jetstream persists the messages to file. It makes our system more resilient and easier to design when our services could restart without losing data

### PR DESCRIPTION
**Pull request by b3rtram**
### Required for all PRs:

- [ x] Updated associated README.md.
- [ ] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

resolves #11046 

Changed nats_consumer input plugin. Added a new configuration element js_subjects where a user can add the subjects to subscribe with jetstream context. If the length of the js_subjects_array is zero it will ignored. js_subject is optional, the original subject is not optional, so there are no breaking changes.
Because there is no interface in go nats consumer I could not reuse nats.Conn, I need to add a new jsConn variable and copy code :(

This is my first github code contribution. Please be patient and give me feedback

### CODE REVIEW

File: plugins/inputs/nats_consumer/README.md [hunk 1]
```diff
@@ -17,6 +17,9 @@ instances of telegraf can read from a NATS cluster in parallel.
   ## subject(s) to consume
   subjects = ["telegraf"]
 
+  ## jetstream subjects
+  js_subjects = ["js_telegraf"]
```
Comment (srebhan) [inline plugins/inputs/nats_consumer/README.md:21]: Can you please make this `jetstream_subjects` to not make the user search for abbreviations!?

File: plugins/inputs/nats_consumer/nats_consumer.go [hunk 2]
```diff
@@ -46,6 +46,7 @@ type natsConsumer struct {
 	Username    string   `toml:"username"`
 	Password    string   `toml:"password"`
 	Credentials string   `toml:"credentials"`
+	JsSubjects  []string `toml:"js_subjects"`
```
Comment (srebhan) [inline plugins/inputs/nats_consumer/nats_consumer.go:49]: With the change above you should do
```suggestion
	JsSubjects  []string `toml:"jetstream_subjects"`
```

File: plugins/inputs/nats_consumer/nats_consumer.go [hunk 3]
```diff
@@ -142,6 +145,31 @@ func (n *natsConsumer) Start(acc telegraf.Accumulator) error {
...
+
+				// set the subscription pending limits
+				err = sub.SetPendingLimits(n.PendingMessageLimit, n.PendingBytesLimit)
+				if err != nil {
+					return err
+				}
+
+				n.jsSubs = append(n.jsSubs, sub)
+			}
+		}
```
Comment (srebhan) [inline plugins/inputs/nats_consumer/nats_consumer.go:172]: Shouldn't this be guarded with a `if n.jsConn == nil || n.jsConn.IsClosed() {` statement similar to `n.conn`?

File: plugins/inputs/nats_consumer/nats_consumer.go [hunk 4]
```diff
@@ -205,6 +233,13 @@ func (n *natsConsumer) clean() {
 		}
 	}
 
+	for _, sub := range n.jsSubs {
+		if err := sub.Unsubscribe(); err != nil {
+			n.Log.Errorf("Error unsubscribing from subject %s in queue %s: %s",
+				sub.Subject, sub.Queue, err.Error())
```
Comment (srebhan) [inline plugins/inputs/nats_consumer/nats_consumer.go:239]: You can use
```suggestion
			n.Log.Errorf("Error unsubscribing from subject %s in queue %s: %v", sub.Subject, sub.Queue, err)
```
which is a bit shorter...

File: plugins/inputs/nats_consumer/sample.conf [hunk 5]
```diff
@@ -6,6 +6,9 @@
   ## subject(s) to consume
   subjects = ["telegraf"]
 
+  ## jetstream subjects
+  js_subjects = ["js_telegraf"]
+
```
Comment (srebhan) [inline plugins/inputs/nats_consumer/sample.conf:11]: Can you please make this `jetstream_subjects` to not make the user search for abbreviations!?

File: telegraf.config [hunk 6]
```diff
@@ -0,0 +1,131 @@
+# Telegraf Configuration
+#
+# Telegraf is entirely plugin driven. All metrics are gathered from the
```
Comment (srebhan) [inline telegraf.config:3]: You don't need to update this file as it is auto-generated IIRC.

**Review by srebhan — COMMENTED** [review]
Hey @b3rtram and thanks for the nice contribution! I have some smaller comments in the code. Can you please take a look?

File: plugins/inputs/nats_consumer/nats_consumer.go (hunk 3 above)
Comment (b3rtram) [reply plugins/inputs/nats_consumer/nats_consumer.go]: n.jsConn.IsClosed is not availabe but I adden if n.jsConn == nil

File: plugins/inputs/nats_consumer/README.md [hunk 7]
```diff
@@ -17,6 +17,9 @@ instances of telegraf can read from a NATS cluster in parallel.
   ## subject(s) to consume
   subjects = ["telegraf"]
 
+  ## jetstream subjects
+  jetstream_subjects = ["js_telegraf"]
```
Comment (srebhan) [inline plugins/inputs/nats_consumer/README.md:21]: Can you maybe add some more words on what this option does and where we can get an introduction to what `jetstream`s are? I doesn't have to be in the "config" section, but some more explanation will help new users to find their way...

**Review by srebhan — APPROVED** [review]
The code looks good to me. I'd nevertheless like to ask you for a few more words in the README to help new/inexperienced users to find their way...

File: plugins/inputs/nats_consumer/README.md (hunk 7 above)
Comment (b3rtram) [reply plugins/inputs/nats_consumer/README.md]: I have added some more information to sample.config and README.md

**Review by srebhan — APPROVED** [review]
Perfect! Thanks @b3rtram!

File: plugins/inputs/nats_consumer/nats_consumer.go [hunk 8]
```diff
@@ -142,6 +145,33 @@ func (n *natsConsumer) Start(acc telegraf.Accumulator) error {
...
 		}
+
+		if len(n.JsSubjects) > 0 {
+			var connErr error
+			n.jsConn, connErr = n.conn.JetStream(nats.PublishAsyncMaxPending(256))
+			if connErr != nil {
+				return connErr
+			}
+
+			if n.jsConn == nil {
```
Comment (sspaink) [inline plugins/inputs/nats_consumer/nats_consumer.go:156]: Shouldn't this be checking if `n.jsConn` isn't nil before continuing? Wouldn't `n.jsConn.QueueSubscribe` fail if it **is** nill?

```suggestion
			if n.jsConn != nil {
```

Comment (b3rtram) [reply plugins/inputs/nats_consumer/nats_consumer.go]: you are right, thank you. have changed it

**Review by srebhan — APPROVED** [review]

**Review by powersj — APPROVED** [review]

**Pull request merged**